scale/scheduler/scheduling/manager.py

```diff
@@ -336,94 +336,122 @@ def _process_queue(self, nodes, job_types, job_type_limits, job_type_resources,
+                    description = INSUFFICIENT_RESOURCES.description % insufficient_resources
+                    scheduler_mgr.warning_active(warning, description)
+
+                if invalid_resources or insufficient_resources:
+                    invalid_resources.extend(insufficient_resources)
+                    jt.unmet_resources = ','.join(invalid_resources)
+                    jt.save(update_fields=["unmet_resources"])
+                    continue
+                else:
+                    # reset unmet_resources flag
+                    jt.unmet_resources = None
+                    scheduler_mgr.warning_inactive(warning)
+                    jt.save(update_fields=["unmet_resources"])
+
+                # Make sure execution's job type and workspaces have been synced to the scheduler
+                job_type_id = queue.job_type_id
+                if job_type_id not in job_types:
+                    scheduler_mgr.warning_active(UNKNOWN_JOB_TYPE, description=UNKNOWN_JOB_TYPE.description % job_type_id)
+                    continue
+
+                workspace_names = job_exe.configuration.get_input_workspace_names()
+                workspace_names.extend(job_exe.configuration.get_output_workspace_names())
+
+                missing_workspace = False
+                for name in workspace_names:
+                    missing_workspace = missing_workspace or name not in workspaces
+                if missing_workspace:
+                    if jt.name in type_warnings:
+                        type_warnings[jt.name]['count'] += 1
+                    else:
+                        type_warnings[jt.name] = {
+                            'warning': '%s job types could not be scheduled due to missing workspace' % jt.name,
+                            'count': 1
+                        }
+                    continue
+
+                # Check limit for this execution's job type
+                if job_type_id in job_type_limits and job_type_limits[job_type_id] < 1:
+                    if jt.name in type_warnings:
+                        type_warnings[jt.name]['count'] += 1
+                    else:
+                        type_warnings[jt.name] = {
+                            'warning': '%s job types could not be scheduled due to scheduling limit reached' % jt.name,
+                            'count': 1
+                        }
+                    continue
+
+                # Try to schedule job execution and adjust job type limit if needed
+                if self._schedule_new_job_exe(job_exe, nodes, job_type_resources):
+                    scheduled_job_executions.append(job_exe)
+                    if job_type_id in job_type_limits:
+                        job_type_limits[job_type_id] -= 1
+        else:
+            logger.warning('There are no nodes available. Waiting to schedule until there are free resources...')
 
         duration = now() - started
+        if type_warnings:
```

Comment:
This will still spam on each pass of the scheduler loop. This might be enough for now, but in the future we might want to consider only outputting warnings once every 5 seconds or some such.

Reply:
It at least won't spam 450 lines out of the 500 loops (or whatever the queue limit is set to). Just one line for each job type after it loops through the queue to give you a summary of what jobs couldn't be scheduled that time around.